### PR TITLE
add data migration to migrate to core software models

### PR DESCRIPTION
**Pull request by gsnider2195**
Blocked by https://github.com/nautobot/nautobot/issues/5280 but this is almost finished and the review should be started now

The only functional change introduced by the core model changes is moving the `default_image` from the `SoftwareImageLCM` model (indicating a default image for the related software version) to the `DeviceType.software_image_files` through table (indicating a default image for this software version for this device type)

I opted to migrate the related objects (like `ObjectChange` and `RelationshipAssociation` instances) instead of creating a new copy of those instances to point to the new models. Since the DLM models will be going away I don't see any reason to keep a copy of that old data around. Does anybody see any downsides to that?

### CODE REVIEW

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (gsnider2195) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:250]: Should probably output a message here if the mapping fails

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:217]: Would it make sense to skip over the relationships that we know are replaced by true FK/M2M fields in the core model, or is that unnecessary optimization here?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:163]: Wonder if it would make sense to make this function a part of core instead - seems pretty generic to any case where we're migrating models from another app?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:246]: Maybe add some comments here for the benefit of those less familiar with `difflib`?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:52]:
```suggestion
            status=status_active,  # DLM model lacks a status field so we default to active
```

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:160]: Do we want to delete the old DLM objects when done, or is that out of scope for now?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:100]: no `object_data_v2`?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:95]: would ACTION_CREATE make more sense?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:104]: Consider having a single `request_id` shared across all ObjectChanges from this migration, instead of a different one for each?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:97]: Interestingly, although `ObjectChange.user` is nullable, `ObjectChange.user_name` is not `blank=True`. Maybe an oversight in core?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:99]: Consider adding the origin DLM object as the `related_object_id`/`related_object_type` on this ObjectChange?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:122]: Comparing to the previous code, no `refresh_from_db()` here?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (timizuoebideri1) [inline nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:14]: Maybe breakup function, looks quite large?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (gsnider2195) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:217]: I was thinking we might want to delete that relationship as part of this migration since leaving it there makes the create device UI confusing

image

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (gsnider2195) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:160]: I think we should delete the models before we release this migration in a DLM release but we might want to hold off until the rest of the DLM changes are being worked on? We'll want to make sure there aren't any other apps/integrations using these models before we delete them and I would think that the next round of DLM updates will be more focused on that.

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (gsnider2195) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:100]: I think we might have had problems doing the DRF serialization in migrations since we're not using the real models? I can double check to see if it looks like it will work here.

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (gsnider2195) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:95]: I was planning to preserve the change history for the old models by changing the `ObjectChange.changed_object_type` content type from the old model to the new model and reusing the primary key. In that case, I think update is more suitable?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:160]: Yeah I don't think we want to delete the model classes immediately, but it might be reasonable/desirable to delete all instances (records) of these models?

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (gsnider2195) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:100]: Looks like I was skipping it in other migrations because we're not using the real models so the DRF serializer can't populate the fields properly. It looks like it's just the display and natural_slug fields that are affected here. Do you think we should use it as is?

```json
{
    "id": "4642a1f7-4bf1-49fc-ac75-66df0ddc02b7",
    "url": "/api/dcim/software-image-files/4642a1f7-4bf1-49fc-ac75-66df0ddc02b7/",
    "tags": [
        {
            "id": "a4b16991-191f-4940-95b4-1b5dd13e19f5",
            "url": "/api/extras/tags/a4b16991-191f-4940-95b4-1b5dd13e19f5/",
            "name": "Amber",
            "color": "ffc107",
            "created": "2023-11-09T16:00:00-08:00",
            "display": "Amber",
            "notes_url": "/api/extras/tags/a4b16991-191f-4940-95b4-1b5dd13e19f5/notes/",
            "description": "",
            "object_type": "extras.tag",
            "last_updated": "2023-11-10T06:59:44.936910-08:00",
            "natural_slug": "amber_a4b1",
            "custom_fields": {}
        }
    ],
    "status": {
        "id": "4e4c9239-7e70-4632-98aa-311090a61824",
        "url": "/api/extras/statuses/4e4c9239-7e70-4632-98aa-311090a61824/",
        "name": "Active",
        "color": "4caf50",
        "created": "2023-11-09T16:00:00-08:00",
        "display": "Status object (4e4c9239-7e70-4632-98aa-311090a61824)",
        "notes_url": "/api/extras/statuses/4e4c9239-7e70-4632-98aa-311090a61824/notes/",
        "description": "Unit is active",
        "object_type": "extras.status",
        "last_updated": "2023-11-10T06:57:23.366251-08:00",
        "natural_slug": "unknown",
        "custom_fields": {}
    },
    "created": "2024-02-20T15:18:46.750295-08:00",
    "display": "SoftwareImageFile object (4642a1f7-4bf1-49fc-ac75-66df0ddc02b7)",
    "notes_url": "/api/dcim/software-image-files/4642a1f7-4bf1-49fc-ac75-66df0ddc02b7/notes/",
    "object_type": "dcim.softwareimagefile",
    "download_url": "",
    "last_updated": "2024-02-21T12:08:43.505074-08:00",
    "natural_slug": "unknown",
    "custom_fields": {
        "custom_field_for_dlm_software": "hp_object_tags"
    },
    "default_image": false,
    "image_file_name": "test_hp_object_tags.bin",
    "image_file_size": null,
    "software_version": {
        "id": "13cd62e1-21e6-4ccc-ac6d-47e89cf6077a",
        "url": "/api/dcim/software-versions/13cd62e1-21e6-4ccc-ac6d-47e89cf6077a/",
        "alias": "device types",
        "status": {
            "id": "4e4c9239-7e70-4632-98aa-311090a61824",
            "url": "/api/extras/statuses/4e4c9239-7e70-4632-98aa-311090a61824/",
            "object_type": "extras.status"
        },
        "created": "2024-02-20T15:00:11.543075-08:00",
        "display": "SoftwareVersion object (13cd62e1-21e6-4ccc-ac6d-47e89cf6077a)",
        "version": "software version device types",
        "platform": {
            "id": "77ec06de-68a6-4b50-a4b3-24283ea111e3",
            "url": "/api/dcim/platforms/77ec06de-68a6-4b50-a4b3-24283ea111e3/",
            "object_type": "dcim.platform"
        },
        "notes_url": "/api/dcim/software-versions/13cd62e1-21e6-4ccc-ac6d-47e89cf6077a/notes/",
        "object_type": "dcim.softwareversion",
        "pre_release": false,
        "last_updated": "2024-02-21T12:08:43.402080-08:00",
        "natural_slug": "unknown",
        "release_date": null,
        "custom_fields": {
            "custom_field_for_dlm_software": "custom field for software version device types"
        },
        "documentation_url": "",
        "long_term_support": false,
        "end_of_support_date": null
    },
    "hashing_algorithm": "",
    "image_file_checksum": ""
}
```

Comment (glennmatthews) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:100]: natural_slug and display are pretty irrelevant for change logging, I think that's close enough.

File: nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py
Comment (glennmatthews) [reply nautobot_device_lifecycle_mgmt/migrations/0021_migrate_to_core_software_models.py:97]: ObjectChange doesn't have any forms, true. So I think you're right.

**Review by HanlinMiao — APPROVED** [review]
LGTM!

**Comment by gsnider2195** [comment]
@progala @jvanderaa would you like to review this before I merge it?

**Comment by jvanderaa** [comment]
@gsnider2195 what's with the tests failing?

**Pull request merged**
